File: src/frontend/lib/api/distributor.ts
import request from './request';
import {
	Distributor,
	DistributorBoxOffice,
	DistributorFilmsData,
	DistributorListData,
} from '@/interfaces/Distributor';
import {
	getDistributorBoxOfficeEndpoint as fetchDistributorBoxOfficeEndpoint,
	getDistributorEndpoint,
	getDistributorFilmsEndpoint,
	getDistributorListEndpoint,
} from './endpoints';
import { FilmSortOption } from '@/interfaces/Film';
import { get, getBoxOffice, getFilms, list } from '@/db/distributor';

/**
 * Distributors
 */
/**
 * Get paginated list of all distributors.
 * @param {number} page - Page number to start from.
 * @param {number} limit - Number of items per page.
 * @returns paginated list of distributors from the api.
 * @example
 * const data = await fetchDistributors(1, 10);
 */
export const fetchDistributors = async (
	page: number = 1,
	limit: number = 10
): Promise<DistributorListData> => {
	if (process.env.USE_PRISMA) {
		try {
			const distributors = await list(page, limit);
			if (list === null) {
				throw new Error();
			}
			return distributors;
		} catch (error) {
			console.warn(error);
			return {
				count: 0,
				next: 0,
				previous: 0,
				results: [],
			};
		}
	}
	try {
		const url = getDistributorListEndpoint(page, limit);
		return await request<DistributorListData>(url, {
			next: { revalidate: 60 },
		});
	} catch (error) {
		console.warn(error);
		return {
			count: 0,
			next: 0,
			previous: 0,
			results: [],
		};
	}
};

/**
 * Get a single distributor.
 * @param {string} slug - Distributor slug.
 * @returns a single distributor from the api.
 * @example
 * const distributor = await fetchDistributor('warner-bros');
 */
export async function fetchDistributor(
	slug: string
): Promise<Distributor | undefined> {
	if (process.env.USE_PRISMA) {
		try {
			const country = await get(slug);
			if (country === null) {
				throw new Error(`distributor with slug '${slug}' not found.`);
			}
			return country;
		} catch (error) {
			console.warn(error);
			return;
		}
	}
	try {
		const url = getDistributorEndpoint(slug);
		return await request<Distributor>(url, { next: { revalidate: 60 } });
	} catch (error) {
		console.warn(error);
		return;
	}
}

/**
 * Gets a distributor box office grouped by year
 * @param {string} slug - Distributor slug.
 * @param {number} limit - Years to go back .
 * @returns a distributors box office grouped by year
 * @example
 * const distributor = await fetchDistributorBoxOffice('warner-bros');
 */
export async function fetchDistributorBoxOffice(
	slug: string,
	limit: number = 25
): Promise<DistributorBoxOffice | undefined> {
	if (process.env.USE_PRISMA) {
		try {
			const results = await getBoxOffice(slug, limit);
			if (results === null) {
				throw new Error();
			}
			return results;
		} catch (error) {
			console.warn(error);
			return;
		}
	}
	try {
		const url = fetchDistributorBoxOfficeEndpoint(slug, limit);
		return await request<DistributorBoxOffice>(url, {
			next: { revalidate: 60 },
		});
	} catch (error) {
		console.warn(error);
		return;
	}
}

/**
 * Get a single distributor films.
 * @param {string} slug - Distributor slug.
 * @param {number} page - Page number to start from.
 * @param {number} limit - Number of items per page.
 * @returns a single distributor and its films from the api.
 * @example
 * const data = fetchDistributorFilms('mubi', 1, 10);
 */
export const fetchDistributorFilms = async (
	slug: string,
	page: number = 1,
	limit: number = 10,
	sort: FilmSortOption
): Promise<DistributorFilmsData | undefined> => {
	if (process.env.USE_PRISMA) {
		try {
			const films = await getFilms(slug, page, limit, sort);
			if (films === null) {
				throw new Error(`Not found`);
			}
			return films;
		} catch (error) {
			console.warn(error);
			return;
		}
	}
	try {
		const url = getDistributorFilmsEndpoint(slug, page, limit, sort);
		return await request<DistributorFilmsData>(url, {
			next: { revalidate: 60 },
		});
	} catch (error) {
		console.warn(error);
		return;
	}
};

File: src/backend/ukbo/services/distributor.py
import datetime
from typing import List, Optional

import pandas as pd
from flask import Response, abort, jsonify
from slugify import slugify  # type: ignore
from sqlalchemy.sql import func
from ukbo import models, services
from ukbo.dto import DistributorSchema, FilmSchemaStrict
from ukbo.extensions import db


def list(page: int = 1, limit: int = 100) -> Response:
    """
    Paginated list of all distributors.

    Args:
        page: Page number
        limit: Number of results per page

    Returns (JSON): Paginated list of distributors.
    """
    query = db.session.query(models.Distributor)
    data = query.order_by(models.Distributor.name.asc()).paginate(
        page=page, per_page=limit, error_out=False
    )

    if data is None:
        abort(404)

    next_page = (page + 1) if data.has_next else ""
    previous_page = (page - 1) if data.has_prev else ""

    distributor_schema = DistributorSchema()

    return jsonify(
        count=data.total,
        next=next_page,
        previous=previous_page,
        results=[distributor_schema.dump(ix) for ix in data.items],
    )


def get(slug: str) -> Response:
    """
    Get one distributor based on its slug.

    Args:
        slug: Slug of the distributor to get.

    Returns (JSON): Distributor data.
    """
    query = db.session.query(models.Distributor)
    query = query.filter(models.Distributor.slug == slug)
    data = query.first()

    if data is None:
        abort(404)

    distributor_schema = DistributorSchema()

    return distributor_schema.dump(data)


def get_films(
    slug: str,
    sort_filter: services.filters.SortFilter = services.filters.SortFilter(),
    page: int = 1,
    limit: int = 100,
) -> Response:
    """
    Get a distributor list of films from slug.

    Args:
        slug: Slug of the distributor to get.
        page: Page number
        limit: Number of results per page

    Returns (JSON): Paginated list of films.
    """
    query = db.session.query(models.Distributor)
    query = query.filter(models.Distributor.slug == slug)
    distributor = query.first()

    if distributor is None:
        abort(404)

    query = db.session.query(models.Film).options(
        db.joinedload(models.Film.weeks)
    )
    query = query.join(models.distributors)
    query = query.join(models.Distributor)

    query = query.filter(models.Distributor.slug == slug)

    # Apply sorting
    query = services.filters.apply_filters(query, sort_filter)

    data = query.paginate(page=page, per_page=limit, error_out=False)

    next_page = (page + 1) if data.has_next else ""
    previous_page = (page - 1) if data.has_prev else ""

    distributor_schema = DistributorSchema()
    film_schema = FilmSchemaStrict()

    return jsonify(
        distributor=distributor_schema.dump(distributor),
        count=data.total,
        next=next_page,
        previous=previous_page,
        results=[film_schema.dump(ix) for ix in data.items],
    )


def get_box_office(slug: str, limit: int) -> Response:
    """
    Gets box office summary for a given distributor, grouped by years.

    Args:
        slug: Slug of the distributor to get.
        limit: The number of years to go backwards.

    Returns:
        JSON response of the list of years.
    """

    query = db.session.query(
        func.extract("year", models.Film_Week.date),
        func.sum(models.Film_Week.total_gross),
        func.count(models.Film.id),
    ).group_by(func.extract("year", models.Film_Week.date))

    query = (
        query.join(models.Film, models.Film.id == models.Film_Week.film_id)
        .join(models.distributors)
        .join(models.Distributor)
    )

    query = query.filter(models.Distributor.slug == slug)

    # get current year and set limit
    now = datetime.datetime.now().year
    query = query.filter(
        func.extract("year", models.Film_Week.date) >= (now - limit)
    )

    data = query.order_by(
        func.extract("year", models.Film_Week.date).desc()
    ).all()

    return jsonify(
        results=[
            dict(
                year=row[0],
                total=row[1],
                count=row[2],
            )
            for row in data
        ]
    )


def add_distributor(distributor: str) -> Optional[List[models.Distributor]]:
    """
    Add a distributor to the database.
    If the distributor is a list within a string separated by ``/``
    It is split, and each one mapped to the full distributor name.
    Checks the database if the distributor exists.
    If not - creates it, adds it to the database.

    Args:
        distributor: Distributor to add.

    Returns list of Distributor objects.
    """
    if distributor is None or (
        isinstance(distributor, str) and not distributor.strip()
    ):
        return None

    distributors = [c.strip() for c in distributor.split("/")]
    new_distributors = []

    for name in distributors:
        name = spellcheck_distributor(name)
        slug = slugify(name)
        db_distributor = models.Distributor.query.filter_by(slug=slug).first()

        if db_distributor and slug == db_distributor.slug:
            new_distributors.append(db_distributor)
        else:
            new = models.Distributor.create(name=name, commit=False)
            new_distributors.append(new)

    return new_distributors


def search(search_query: str) -> Response:
    """
    Search distributors by name.

    Args:
        search_query: Search query.

    Returns (JSON): List of distributors.
    """
    query = db.session.query(models.Distributor)
    query = query.filter(models.Distributor.name.ilike(f"%{search_query}%"))
    data = query.limit(10).all()

    distributor_schema = DistributorSchema()

    return [] if data is None else [distributor_schema.dump(ix) for ix in data]


def market_share(year: Optional[str] = None) -> Response:
    """
    Gets the distributor market share for a year

    Args:
        year: Year to get the market share for.

    Returns (JSON): List of distributors and their market share.
    """
    query = db.session.query(
        func.extract("year", models.Film_Week.date),
        models.Distributor,
        func.sum(models.Film_Week.week_gross),
    )

    query = query.join(models.Film)
    query = query.join(models.distributors)
    query = query.join(models.Distributor)
    query = query.group_by(models.Distributor)
    query = query.group_by(func.extract("year", models.Film_Week.date))
    query = query.order_by(func.extract("year", models.Film_Week.date).desc())

    if year is not None:
        query = query.filter(
            func.extract("year", models.Film_Week.date) == year
        )
    else:
        query = query.filter(
            models.Film_Week.date >= datetime.date(2018, 1, 1)
        )

    data = query.all()

    if data is None:
        abort(404)

    distributor_schema = DistributorSchema()

    return jsonify(
        results=[
            dict(
                year=row[0],
                distributor=distributor_schema.dump(row[1]),
                gross=row[2],
            )
            for row in data
        ]
    )


def market_share_date(start: str, end: str) -> Response:
    """
    Gets the distributor market share for a time period.

    Args:
        start: Start date.
        end: End date.

    Returns (JSON): List of distributors and their market share.
    """
    query = db.session.query(
        models.Distributor,
        func.sum(models.Film_Week.week_gross),
    )
    query = query.join(models.Film)
    query = query.join(models.distributors)
    query = query.join(models.Distributor)
    query = query.filter(models.Film_Week.date >= start)
    query = query.filter(models.Film_Week.date <= end)
    query = query.group_by(models.Distributor)
    query = query.order_by(func.sum(models.Film_Week.week_gross).desc())

    data = query.all()

    if data is None:
        abort(404)

    distributor_schema = DistributorSchema()

    return jsonify(
        results=[
            dict(
                distributor=distributor_schema.dump(row[0]),
                gross=row[1],
            )
            for row in data
        ]
    )


def spellcheck_distributor(distributor: pd.Series) -> str:
    """
    Spellchecks the distributor against a list of common mistakes

    Args:
        distributor: Distributor to check

    Returns:
        str: Cleaned distributor
    """
    distributor = distributor.strip().upper()
    try:
        df = pd.read_csv("./data/distributor_check.csv", header=None)
    except FileNotFoundError:
        return distributor
    df.columns = ["key", "correction"]

    if distributor in df["key"].values:
        df = df[df["key"].str.match(distributor)]
        distributor = df["correction"].iloc[0]
    return distributor
